IsNegtive exist big bug

When using roaring64.BSI, due to value type is *big.Int , bsi's bitCount is Variable length even if some negative value is set. Thus there is a bug, when bsi's min negative value's bitLen is less than max positive num, GetValue will get wrong error.

1. unit test will pass

`func TestGetNegativeValue(t *testing.T) {`
	`num := int64(14)`
	`bsi := NewDefaultBSI()`
	`for i:=-num; i<=num;i++ {`
	`	bsi.SetValue(uint64(i+num), i)`
	`}`
	`val, _ := bsi.GetValue(0)`
	`assert.Equal(t, -num, val)`
	`bsi.SetValue(uint64((num+1)+num), num+1)`
	`val, _ = bsi.GetValue(0)`
	`assert.Equal(t, -num, val)`
`}`

2. unit test will fail,  when set value 16 to bsi, bsi's higthest bit is not sign bit
`func TestGetNegativeValue(t *testing.T) {`
	`num := int64(15)`
	`bsi := NewDefaultBSI()`
	`for i:=-num; i<=num;i++ {`
	`	bsi.SetValue(uint64(i+num), i)`
	`}`
	`val, _ := bsi.GetValue(0)`
	`assert.Equal(t, -num, val)`
	`bsi.SetValue(uint64((num+1)+num), num+1)`
	`val, _ = bsi.GetValue(0)`
	`assert.Equal(t, -num, val)`
`}`
Error:      	Not equal: 
        	            	expected: -15
        	            	actual  : 17
        	Test:       	TestGetNegativeValue
--- FAIL: TestGetNegativeValue (0.00s)

Are you attempting to set columnId to a negative value?

On Wed, Nov 27, 2024 at 8:02 AM ChicChip ***@***.***> wrote:

>
>    1. unit test will pass
>
> func TestGetNegativeValue(t *testing.T) {
> num := int64(14)
> bsi := NewDefaultBSI()
> for i:=-num; i<=num;i++ {
> bsi.SetValue(uint64(i+num), i)
> }
> val, _ := bsi.GetValue(0)
> assert.Equal(t, -num, val)
> bsi.SetValue(uint64((num+1)+num), num+1)
> val, _ = bsi.GetValue(0)
> assert.Equal(t, -num, val)
> }
>
>    2. unit test will fail, when set value 16 to bsi, bsi's higthest bit
>    is not sign bit
>    func TestGetNegativeValue(t *testing.T) {
>    num := int64(15)
>    bsi := NewDefaultBSI()
>    for i:=-num; i<=num;i++ {
>    bsi.SetValue(uint64(i+num), i)
>    }
>    val, _ := bsi.GetValue(0)
>    assert.Equal(t, -num, val)
>    bsi.SetValue(uint64((num+1)+num), num+1)
>    val, _ = bsi.GetValue(0)
>    assert.Equal(t, -num, val)
>    }
>    Error: Not equal:
>    expected: -15
>    actual : 17
>    Test: TestGetNegativeValue
>    --- FAIL: TestGetNegativeValue (0.00s)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/RoaringBitmap/roaring/issues/462#issuecomment-2503958263>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZZAUTPTVSHT53FOZJRTDD2CXGHNAVCNFSM6AAAAABSSM5VLKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBTHE2TQMRWGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

--- a/roaring64/bsi64.go
+++ b/roaring64/bsi64.go
@@ -100,8 +100,20 @@ func (b *BSI) SetBigValue(columnID uint64, value *big.Int) {
 		if minBits == 1 {
 			minBits = 2
 		}
-		for len(b.bA) < minBits {
-			b.bA = append(b.bA, Bitmap{})
+		if len(b.bA) < minBits {
+			oldSignPos := len(b.bA) - 1
+			for len(b.bA) < minBits {
+				b.bA = append(b.bA, Bitmap{})
+			}
+			// When bA grows, the sign slot shifts from oldSignPos to the new end
+			// of bA. For existing negative entries (whose sign bit is set in
+			// bA[oldSignPos]), sign-extension requires that all intermediate bit
+			// positions between oldSignPos and the new sign position also be set.
+			// Copy the old sign bitmap into every new slot (sign extension).
+			newSignPos := len(b.bA) - 1
+			for i := oldSignPos + 1; i <= newSignPos; i++ {
+				b.bA[i].Or(&b.bA[oldSignPos])
+			}
 		}
 	}
 
@@ -122,8 +134,16 @@ func (b *BSI) SetBigMany(foundSet *Bitmap, value *big.Int) {
 		if minBits == 1 {
 			minBits = 2
 		}
-		for len(b.bA) < minBits {
-			b.bA = append(b.bA, Bitmap{})
+		if len(b.bA) < minBits {
+			oldSignPos := len(b.bA) - 1
+			for len(b.bA) < minBits {
+				b.bA = append(b.bA, Bitmap{})
+			}
+			// Sign-extend existing negative entries into the new bit slots.
+			newSignPos := len(b.bA) - 1
+			for i := oldSignPos + 1; i <= newSignPos; i++ {
+				b.bA[i].Or(&b.bA[oldSignPos])
+			}
 		}
 	}
 	for i := b.BitCount(); i >= 0; i-- {

--- a/roaring64/bsi64_test.go
+++ b/roaring64/bsi64_test.go
@@ -870,3 +870,26 @@ func BenchmarkClearValues(b *testing.B) {
 		b2.ClearValues(resultA)
 	}
 }
+
+// TestGetValueNegativeAfterGrowth checks that negative values stored in a
+// NewDefaultBSI are returned correctly after bA grows because a larger positive
+// value is added later. Previously, the sign bit was stranded at the old slot
+// when bA was extended, causing wrong-sign results.
+// See https://github.com/RoaringBitmap/roaring/issues/462
+func TestGetValueNegativeAfterGrowth(t *testing.T) {
+	bsi := NewDefaultBSI()
+	for i := int64(-15); i <= 15; i++ {
+		bsi.SetValue(uint64(i+15), i)
+	}
+
+	val, exists := bsi.GetValue(0)
+	require.True(t, exists)
+	assert.Equal(t, int64(-15), val, "before growth")
+
+	// Adding 16 requires an extra bit and grows bA, shifting the sign slot.
+	bsi.SetValue(uint64(16+15), 16)
+
+	val, exists = bsi.GetValue(0)
+	require.True(t, exists)
+	assert.Equal(t, int64(-15), val, "after growth")
+}

